### PR TITLE
Do not create health check CR until karpenter.sh/initialized=true

### PR DESCRIPTION
**Pull request by zman-ms**
Do not create health check CR until karpenter.sh/initialized=true, otherwise result is not reliable

### CODE REVIEW

File: pkg/controller/node/controller.go [hunk 1]
```diff
@@ -57,6 +57,13 @@ const (
 	// forever. The value is aligned with Karpenter's Node Auto Repair unready
 	// timeout (10m); the in-house Remediator uses 5m for the same condition.
 	NodeReadyMaxWait = 10 * time.Minute
+
+	// KarpenterUnregisteredTaintKey is the taint key Karpenter applies to
+	// nodes it has provisioned but not yet finished registering. The taint is
+	// removed by Karpenter once the node has been registered, so we defer
+	// CheckNodeHealth creation until then to avoid running checks against a
+	// node that is not yet fully initialized.
+	KarpenterUnregisteredTaintKey = "karpenter.sh/unregistered"
```
Comment (ThatJuanGuy) [inline pkg/controller/node/controller.go]: Maybe it would be better to wait for the karpenter.sh/registered label? I think removing this taint is probably a requirement for it.

Comment (ThatJuanGuy) [reply pkg/controller/node/controller.go]: It might be better to wait for the label for initialization instead of registered
https://deepwiki.com/kubernetes-sigs/karpenter/4-node-lifecycle-management

File: pkg/controller/node/controller.go [hunk 2]
```diff
@@ -57,6 +57,18 @@ const (
 	// forever. The value is aligned with Karpenter's Node Auto Repair unready
 	// timeout (10m); the in-house Remediator uses 5m for the same condition.
 	NodeReadyMaxWait = 10 * time.Minute
+
+	// KarpenterNodePoolLabel is set by Karpenter on every node it manages and
+	// holds the name of the owning NodePool. Its presence is used to detect
+	// whether a node is Karpenter-managed.
+	KarpenterNodePoolLabel = "karpenter.sh/nodepool"
```
Comment (ThatJuanGuy) [inline pkg/controller/node/controller.go]: There's a special type of karpenter node called standalone nodes that won't have this label. https://teams.microsoft.com/l/message/19:b94e67ab23e44a6c940de9b6ee149c95@thread.skype/1770400038601?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=e121dbfd-0ec1-40ea-8af5-26075f6a731b&parentMessageId=1769549506347&teamName=Azure%20Container%20Compute&channelName=Auto%20SIG%20-%20Karpenter&createdTime=1770400038601

To check if a node is karpenter-managed, a more reliable fixed label to use is `karpenter.sh/capacity-type`. That said, it's probably better to check for any `karpenter.sh/` prefixed label's existence.

Comment (zman-ms) [reply pkg/controller/node/controller.go]: updated

File: pkg/controller/node/controller.go [hunk 3]
```diff
@@ -57,6 +57,18 @@ const (
 	// forever. The value is aligned with Karpenter's Node Auto Repair unready
 	// timeout (10m); the in-house Remediator uses 5m for the same condition.
 	NodeReadyMaxWait = 10 * time.Minute
+
+	// KarpenterNodePoolLabel is set by Karpenter on every node it manages and
+	// holds the name of the owning NodePool. Its presence is used to detect
+	// whether a node is Karpenter-managed.
+	KarpenterNodePoolLabel = "karpenter.sh/capacity-type"
```
Comment (ThatJuanGuy) [inline pkg/controller/node/controller.go]: This variable name should be updated.

Comment (ThatJuanGuy) [reply pkg/controller/node/controller.go]: Also the doc comment on IsKarpenterManaged

File: pkg/controller/node/controller.go [hunk 4]
```diff
@@ -202,12 +219,31 @@ func isNodeReady(node *corev1.Node) bool {
 	return false
 }
 
+// isKarpenterManaged reports whether the node is managed by Karpenter,
+// detected via the presence of the karpenter.sh/nodepool label.
+func isKarpenterManaged(node *corev1.Node) bool {
+	v, ok := node.Labels[KarpenterNodePoolLabel]
+	return ok && v != ""
+}
+
```
Comment (ThatJuanGuy) [inline pkg/controller/node/controller.go:246]: I think this would be more reliable. We don't really care about the value.
```suggestion
func isKarpenterManaged(node *corev1.Node) bool {
	_, ok := node.Labels[KarpenterNodePoolLabel]
	return ok
}

```

File: pkg/controller/node/controller.go
Comment (ThatJuanGuy) [inline pkg/controller/node/controller.go:170]: Nit

```suggestion
// createCheckNodeHealth creates a CheckNodeHealth CR with a deterministic
// name, but only if the node has been initialized. Running the health
// checks against a node that has not finished initializing (e.g., still
```

File: pkg/controller/node/controller.go (hunk 4 above)
Comment (zman-ms) [reply pkg/controller/node/controller.go:246]: empty value itself is weird enough that we should not treat it as normal karpenter node. There is no restriction blocking user adding this to non-karpenter nodes. Validating that can be a separate improvement but this non-empty check is not causing extra costs.

File: pkg/controller/node/controller.go
Comment (zman-ms) [reply pkg/controller/node/controller.go:170]: udpated

File: pkg/controller/node/controller.go [hunk 5]
```diff
@@ -168,6 +182,11 @@ func (r *NodeRebootReconciler) createCheckNodeHealth(ctx context.Context, node *
 		klog.InfoS("Node is not Ready yet, deferring CheckNodeHealth creation", "node", node.Name, "bootID", bootID)
 		return false, nil
 	}
+	if isKarpenterManaged(node) && !isKarpenterInitialized(node) {
```
Comment (ThatJuanGuy) [inline pkg/controller/node/controller.go]: I think it would be cleaner to just have a single `isNodeReady()` function. The function should determine the node type and use that information to make the decision. 

The abstraction for the condition we want to verify is cleaner and we don't have to add cases for karpenter in multiple places.

Comment (ThatJuanGuy) [reply pkg/controller/node/controller.go]: notReadyExceeds() can also be simplified in this PR if we do this. It currently copies the code from IsNotReady() inline which makes it hard to follow.

File: pkg/controller/node/controller_test.go [hunk 6]
```diff
@@ -421,4 +469,114 @@ func TestCreateCheckNodeHealthSkipsWhenNotReady(t *testing.T) {
...
+		}
+
+		cnh := &chmv1alpha1.CheckNodeHealth{}
+		if err := fc.Get(context.Background(), client.ObjectKey{Name: GenerateCNHName("node-1", "boot-aaa")}, cnh); err != nil {
+			t.Errorf("expected CheckNodeHealth to be created: %v", err)
+		}
+	})
+}
+
+func TestKarpenterHelpers(t *testing.T) {
```
Comment (ThatJuanGuy) [inline pkg/controller/node/controller_test.go]: Nit: Would prefer to break this out into specific per-function tests. Would make it a bit easier to understand what the relevant parts of each test case/condition are. Also to reason about what parts of the code have test coverage.

Not nit: If leaving this as a single test, please update the test names. It is not very clear what "managed" means in this case as we don't use the terminology elsewhere. We use "karpenter". Also, as the code is currently written, we don't care about the specific value of the capacity type label, just whether a value is present. There should be unit tests for proper behavior in both those cases.

File: pkg/controller/node/controller_test.go [hunk 7]
```diff
@@ -421,4 +469,114 @@ func TestCreateCheckNodeHealthSkipsWhenNotReady(t *testing.T) {
...
+				t.Errorf("isKarpenterManaged = %v, want %v", got, tc.wantManaged)
+			}
+			if got := isKarpenterInitialized(node); got != tc.wantInitialized {
+				t.Errorf("isKarpenterInitialized = %v, want %v", got, tc.wantInitialized)
+			}
+		})
+	}
+}
+
+func TestNotReadyExceedsKarpenterUninitialized(t *testing.T) {
```
Comment (ThatJuanGuy) [inline pkg/controller/node/controller_test.go]: This test shouldn't really exist imo. As mentioned in an earlier comment, ideally I think the karpenter specific check should be handled in IsNodeReady and we unit test karpenter specific cases there. notReadyExceeds should not have to care about the node type, just not ready status.

File: pkg/controller/node/controller.go (hunk 4 above)
Comment (ThatJuanGuy) [reply pkg/controller/node/controller.go:246]: My take here is this is determining whether the node is karpenter managed, not whther the node is in a proper state or karpenter is behaving properly. Even if karpenter is not behaving properly, we should probably treat this node as if it were a karpenter node because we are reasonably confident it will be one.

Comment (ThatJuanGuy) [reply pkg/controller/node/controller.go:246]: That said, I don't feel too strongly about this because I think the odds of this being empty are near 0.

File: pkg/controller/node/controller.go (hunk 5 above)
Comment (zman-ms) [reply pkg/controller/node/controller.go]: notReadyExceeds() is different. It needs to compare the time stamps instead of just getting ready status, but isNodeReady() is doable

File: pkg/controller/node/controller_test.go (hunk 7 above)
Comment (zman-ms) [reply pkg/controller/node/controller_test.go]: updated

File: pkg/controller/node/controller.go
Comment (ThatJuanGuy) [inline pkg/controller/node/controller.go:275]: This function can be no-op rewritten as something like below. It is currently calling the helper function logic inline instead of through the function right now.

```suggestion
func notReadyExceeds(node *corev1.Node, d time.Duration) bool {
	if !IsNodeReadyForHealthCheck(node) {
	    return time.Since(node.CreationTimestamp.Time) > d
	}
	return false
}
```

File: pkg/controller/node/controller_test.go [hunk 8]
```diff
@@ -384,6 +432,62 @@ func TestRemoveStaleNodeCondition(t *testing.T) {
...
 }
 
+func TestIsNodeReadyForHealthCheck(t *testing.T) {
+	tests := []struct {
+		name string
+		node *corev1.Node
+		want bool
+	}{
+		{
+			name: "non-Karpenter Ready=True — ready",
```
Comment (ThatJuanGuy) [inline pkg/controller/node/controller_test.go]: I assume Ready=True is referring to the condition. Seeing as how the term ready is somewhat overloaded (IsReadyForHealthCheck vs node condition ready), it would be nice to clarify NodeConditionReady=true.

File: pkg/controller/node/controller.go
Comment (zman-ms) [reply pkg/controller/node/controller.go:275]: No it also compares with LastTransitionTime

File: pkg/controller/node/controller.go (hunk 5 above)
Comment (ThatJuanGuy) [reply pkg/controller/node/controller.go]: Added a suggestion comment explaining what I meant a bit clearer hopefully.

Comment (ThatJuanGuy) [reply pkg/controller/node/controller.go]: Oh wait nvm I think I see what you mean. This needs to check last transition time in the non-karpenter case. It doesn't always default to CreationTimestamp.

File: pkg/controller/node/controller_test.go (hunk 6 above)
Comment (zman-ms) [reply pkg/controller/node/controller_test.go]: removed

File: pkg/controller/node/controller.go
Comment (ThatJuanGuy) [reply pkg/controller/node/controller.go:275]: Yeah. I see that now. 

Nit: could maybe be re-written something like below so we can reuse the NodeReadyForHealthCheck helper instead of potentially having to update that multiple places in the future. Then have this function determine which time to return. 
```go
func notReadyExceeds(node *corev1.Node, d time.Duration) bool {
	if !IsNodeReadyForHealthCheck(node) {
	    if isKarpenterManaged {
            return time.Since(node.CreationTimestamp.Time) > d
        }

        // non-karpenter managed node
        for _, c := range node.Status.Conditions {
		    if c.Type != corev1.NodeReady {
			    continue
		    }
		    if c.Status == corev1.ConditionTrue {
			    return false
		    }
		    return time.Since(c.LastTransitionTime.Time) > d
	    }
	    return time.Since(node.CreationTimestamp.Time) > d
	}

	return false
}
``` 

It would probably be nice to have a helper function to find a node condition/the transition timestamp.

File: pkg/controller/node/controller.go (hunk 4 above)
Comment (zman-ms) [reply pkg/controller/node/controller.go:246]: ok let's keep it as is since this is a rare edge case and it works either way.

File: pkg/controller/node/controller_test.go (hunk 8 above)
Comment (zman-ms) [reply pkg/controller/node/controller_test.go]: renamed the tests and removed one unrealistic case

File: pkg/controller/node/controller.go
Comment (zman-ms) [reply pkg/controller/node/controller.go:275]: Ok we can think about this later. You may create a housekeeping work item.

**Review by ThatJuanGuy — APPROVED** [review]
